op-dispute-mon/mon/extract/output_agreement_enricher.go

```diff
@@ -30,54 +32,155 @@ type OutputMetrics interface {
+			}
+		}(i, client)
+	}
+	wg.Wait()
+
+	validResults := make([]outputResult, 0, len(results))
+	foundResults := make([]outputResult, 0, len(results))
+	for idx, result := range results {
+		if result.err != nil {
+			o.log.Error("Failed to fetch output root", "clientIndex", idx, "l2BlockNum", game.L2BlockNumber, "err", result.err)
+			continue
+		}
+
+		validResults = append(validResults, result)
+
+		if result.notFound {
+			o.log.Info("Node has not seen block", "clientIndex", idx, "l2BlockNum", game.L2BlockNumber)
+		} else {
+			foundResults = append(foundResults, result)
 		}
-		return fmt.Errorf("failed to get output at block: %w", err)
 	}
+
+	// If all results were errors, return an error
+	if len(validResults) == 0 {
+		return fmt.Errorf("failed to get output at block: %w", ErrAllNodesUnavailable)
+	}
+
+	// If all remaining nodes returned "not found", we disagree with any claim.
+	if len(foundResults) == 0 {
+		game.AgreeWithClaim = false
+		game.ExpectedRootClaim = common.Hash{}
+		return nil
+	}
+
+	// At least one node returned an output root, record the fetch time.
 	o.metrics.RecordOutputFetchTime(float64(o.clock.Now().Unix()))
-	game.ExpectedRootClaim = common.Hash(output.OutputRoot)
-	rootMatches := game.RootClaim == game.ExpectedRootClaim
-	if !rootMatches {
+
+	// Check for disagreements among nodes.
+	// A disagreement is any of:
+	// - Mixed "found" and "not found" responses.
+	// - Different output roots from nodes that found an output.
+	firstResult := foundResults[0]
+	diverged := len(foundResults) < len(validResults)
+	if !diverged {
+		for _, result := range foundResults[1:] {
+			if result.outputRoot != firstResult.outputRoot {
+				diverged = true
+				break
+			}
+		}
+	}
+
+	if diverged {
+		o.log.Warn("Nodes disagree on output root",
+			"l2BlockNum", game.L2BlockNumber,
+			"firstOutput", firstResult.outputRoot,
+			"found", len(foundResults),
+			"valid", len(validResults))
 		game.AgreeWithClaim = false
+		game.ExpectedRootClaim = firstResult.outputRoot
 		return nil
 	}
 
-	// If the root matches, also check that l2 block is safe at the L1 head
-	safeHead, err := o.client.SafeHeadAtL1Block(ctx, game.L1HeadNum)
-	if err != nil {
-		o.log.Warn("Unable to verify proposed block was safe", "l1HeadNum", game.L1HeadNum, "l2BlockNum", game.L2BlockNumber, "err", err)
-		// If safe head data isn't available, assume the output root was safe
-		// Avoids making the dispute mon dependent on safe head db being available
-		game.AgreeWithClaim = true
+	// All nodes that found an output agree on the root.
+	// Now check if the output is considered safe by at least one node.
+	atLeastOneSafe := false
+	for _, result := range foundResults {
+		if result.isSafe {
+			atLeastOneSafe = true
+			break
+		}
+	}
+
+	// If no node considers the output safe, we disagree.
+	if !atLeastOneSafe {
+		o.log.Warn("All nodes agree on output root, but none consider it safe",
+			"l2BlockNum", game.L2BlockNumber, "root", firstResult.outputRoot)
```

Comment:
I wouldn't warn about this. It's a normal situation that applies to every game from the invalid proposer. Basically we are totally ok with invalid games existing as long as they have been countered by the challenger.  This code is just loading the game state and then other steps in the update cycle will analyse the state and report what's happening via metrics and either warn logs for things that may be a problem (eg game forecast to resolve wrong) or error logs for things that are definitely wrong (game _has_ resolved wrong).